### PR TITLE
Rearm limited ammo truck munition

### PR DESCRIPTION
**Pull request by GitHawk**
The rearm module will be expanded by a new option to select the ammunition amount in an ammo truck.
The two new availability levels are:
- Amount based on caliber
- Certain magazines and specific amounts
  The current "unlimited" level will remain as default value.

Supply amount based on caliber means you can for example take either two .50cal magazines or a single 120mm magazine, but not a 155mm magazine when the amount is on the level of 154. So everything can be rearmed but with a limited supply, while bigger calibers take more supply.

Supply amount with certain magazines means you are required to add certain magazines to an ammo truck. No other weapons/magazines will be able to be resupplied.
- [x] Implement the two levels
- [x] Add API
- [x] Testing

Documentation for other modders/creators on how to enable refuel will be in a different PR

### CODE REVIEW

**Comment by jonpas** [comment]
I'd wait with the module additions until we decide what to do with settings/Eden attributes.

File: addons/rearm/functions/fnc_rearmSuccess.sqf [hunk 1]
```diff
@@ -21,23 +21,21 @@
...
     if (_turretOwnerID == 0) then {
-        [_this, QFUNC(rearmSuccessLocal), _target] call EFUNC(common,execRemoteFnc);
+        [QGVAR(rearmSuccessLocalEH), _vehicle, _this] call EFUNC(common,objectEvent);
     } else {
-        EGVAR(common,remoteFnc) = [_this, QFUNC(rearmSuccessLocal), 0];
-        _turretOwnerID publicVariableClient QEGVAR(common,remoteFnc);
+        [QGVAR(rearmSuccessLocalEH), _turretOwnerID, _this] call EFUNC(common,targetEvent);
     };
 } else {
     [_this, QFUNC(rearmSuccess), 1] call EFUNC(common,execRemoteFnc);
```
Comment (jonpas) [inline addons/rearm/functions/fnc_rearmSuccess.sqf]: As discussed on Slack, use `EFUNC(common,serverEvent)` here and in postInit add the event in `isServer` only.

**Comment by Glowbal** [comment]
Has this been reviewed? It's a significant amount of code.

**Comment by GitHawk** [comment]
No, I tested it a lot while writing it. That was 2 ArmA versions ago :-(

**Comment by jonpas** [comment]
@GitHawk any further testing on this?

File: addons/common/functions/fnc_displayTextStructured.sqf [hunk 2]
```diff
@@ -1,11 +1,12 @@
 /*
- * Author: commy2, Glowbal
+ * Author: commy2, Glowbal, GitHawk
  * Display a structured text.
  *
  * Arguments:
  * 0: Text <ANY>
  * 1: Size of the textbox (default: 1.5) <NUMBER>
  * 2: Target Unit. Will only display if target is the player controlled object (default: ACE_player) <OBJECT>
+ * 3: Custom Width <NUMBER><OPTIONAL>
```
Comment (jonpas) [inline addons/common/functions/fnc_displayTextStructured.sqf]: Why is that necessary? You can just calculate the size based on length of the string, not sure I like this.

File: addons/rearm/CfgVehicles.hpp [hunk 3]
```diff
@@ -14,8 +14,28 @@
         };
 
 #define MACRO_REARM_TRUCK_ACTIONS \
-        class ACE_Actions: ACE_Actions { \
-            class ACE_MainActions: ACE_MainActions { \
+        class ACE_Actions : ACE_Actions { \
+            class ACE_MainActions : ACE_MainActions { \
```
Comment (jonpas) [inline addons/rearm/CfgVehicles.hpp]: I'd leave those 2 lines like there were, consistency.

File: addons/rearm/functions/fnc_addMagazineToSupply.sqf [hunk 4]
```diff
@@ -0,0 +1,68 @@
+/*
+ * Author: GitHawk
+ * Adds magazines to the supply.
+ *
+ * Argument:
+ * 0: Ammo Truck <OBJECT>
+ * 1: Magazine Classname <STRING>
+ * 2: Only partial <BOOL><OPTIONAL>
```
Comment (jonpas) [inline addons/rearm/functions/fnc_addMagazineToSupply.sqf]: This is how we do optional arguments documentation:
```
 * 2: Only partial <BOOL> (default: false)
```

File: addons/common/functions/fnc_displayTextStructured.sqf (hunk 2 above)
Comment (GitHawk) [inline addons/common/functions/fnc_displayTextStructured.sqf]: In its original way, you couldn't manipulate the width at all.So in order to break less, I added an optional parameter.

File: addons/rearm/functions/fnc_addMagazineToSupply.sqf [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+params [
+    ["_truck", objNull, [objNull]],
+    ["_magazineClass", "", [""]],
+    ["_partial", false, [false]]
+];
+
+if (isNull _truck ||
+    {_magazineClass isEqualTo ""} ||
+    {GVAR(supply) == 0}) exitWith {};
```
Comment (jonpas) [inline addons/rearm/functions/fnc_addMagazineToSupply.sqf]: Those would be more readable in single line.

File: addons/rearm/functions/fnc_addVehicleMagazinesToSupply.sqf [hunk 6]
```diff
@@ -0,0 +1,47 @@
+/*
+ * Author: GitHawk
+ * Adds all magazines of a vehicle to the supply.
+ *
+ * Argument:
+ * 0: Ammo Truck <OBJECT>
+ * 1: Vehicle <OBJECT>
+ * or
+ * 0: Ammo Truck <OBJECT>
+ * 1: Vehicle class <STRING>
```
Comment (jonpas) [inline addons/rearm/functions/fnc_addVehicleMagazinesToSupply.sqf]:
```
 * 0: Ammo Truck <OBJECT>
 * 1: Vehicle or Vehicle class <OBJECT/STRING>
```

File: addons/rearm/functions/fnc_addVehicleMagazinesToSupply.sqf [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+ *
+ * Return value:
```
Comment (jonpas) [inline addons/rearm/functions/fnc_addVehicleMagazinesToSupply.sqf]: **V**alue

File: addons/rearm/functions/fnc_addVehicleMagazinesToSupply.sqf [hunk 8]
```diff
@@ -0,0 +1,47 @@
+/*
+ * Author: GitHawk
+ * Adds all magazines of a vehicle to the supply.
+ *
+ * Argument:
```
Comment (jonpas) [inline addons/rearm/functions/fnc_addVehicleMagazinesToSupply.sqf]: Argument**s**

File: addons/rearm/functions/fnc_addMagazineToSupply.sqf [hunk 9]
```diff
@@ -0,0 +1,68 @@
+/*
+ * Author: GitHawk
+ * Adds magazines to the supply.
+ *
+ * Argument:
+ * 0: Ammo Truck <OBJECT>
+ * 1: Magazine Classname <STRING>
+ * 2: Only partial <BOOL><OPTIONAL>
+ *
+ * Return value:
```
Comment (jonpas) [inline addons/rearm/functions/fnc_addMagazineToSupply.sqf]: **V**alue

File: addons/rearm/functions/fnc_addMagazineToSupply.sqf [hunk 10]
```diff
@@ -0,0 +1,68 @@
+/*
+ * Author: GitHawk
+ * Adds magazines to the supply.
+ *
+ * Argument:
```
Comment (jonpas) [inline addons/rearm/functions/fnc_addMagazineToSupply.sqf]: Argument**s**

File: addons/rearm/functions/fnc_addVehicleMagazinesToSupply.sqf [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+ */
+#include "script_component.hpp"
+
+params [
+    ["_truck", objNull, [objNull]],
+    ["_vehicle", objNull, [objNull, ""]]
+];
+
+if (isNull _truck ||
+    {typeName _vehicle == "OBJECT" && {isNull _vehicle}}) exitWith {};
```
Comment (jonpas) [inline addons/rearm/functions/fnc_addVehicleMagazinesToSupply.sqf]: Same line looks better. And use `_vehicle isEqualType objNull`.

File: addons/rearm/functions/fnc_addVehicleMagazinesToSupply.sqf [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+
+private _string = "";
+if (typeName _vehicle == "OBJECT") then {
+    _string = typeOf _vehicle;
+};
+if (typeName _vehicle == "STRING") then {
+    _string = _vehicle;
+};
```
Comment (jonpas) [inline addons/rearm/functions/fnc_addVehicleMagazinesToSupply.sqf]: L31-L37: `private _string = [_vehicle, typeOf _vehicle] select (_vehicle isEqualType objNull);`

File: addons/rearm/functions/fnc_canReadSupplyCounter.sqf [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+    ["_truck", objNull, [objNull]],
+    ["_unit", objNull, [objNull]]
+];
+
+!(isNull _unit ||
+    {!(_unit isKindOf "CAManBase")} ||
+    {!local _unit} ||
+    {!alive _truck} ||
+    {(_truck distance _unit) > REARM_ACTION_DISTANCE} ||
+    {GVAR(supply) == 0})
```
Comment (jonpas) [inline addons/rearm/functions/fnc_canReadSupplyCounter.sqf]: L25-L29 doesn't need indentation.

File: addons/rearm/functions/fnc_canRearmVehicle.sqf [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+    ["_vehicle", objNull, [objNull]],
+    ["_unit", objNull, [objNull]]
+];
+
+!(isNull _unit ||
+    {!(_unit isKindOf "CAManBase")} ||
+    {!local _unit} ||
+    {(_vehicle distance _unit) > REARM_ACTION_DISTANCE} ||
+    {!isNull (_unit getVariable [QGVAR(dummy), objNull])} ||
+    {GVAR(level) > 0})
```
Comment (jonpas) [inline addons/rearm/functions/fnc_canRearmVehicle.sqf]: No indentation needed either.

File: addons/rearm/functions/fnc_canStoreAmmo.sqf [hunk 15]
```diff
@@ -10,16 +10,19 @@
...
+    ["_truck", objNull, [objNull]],
+    ["_unit", objNull, [objNull]]
+];
 
 !(isNull _unit ||
     {!(_unit isKindOf "CAManBase")} ||
     {!local _unit} ||
-    {(_target distance _unit) > REARM_ACTION_DISTANCE} ||
+    {(_truck distance _unit) > REARM_ACTION_DISTANCE} ||
     {isNull (_unit getVariable [QGVAR(dummy), objNull])})
```
Comment (jonpas) [inline addons/rearm/functions/fnc_canStoreAmmo.sqf]: No indentation needed either.

File: addons/rearm/functions/fnc_canTakeAmmo.sqf [hunk 16]
```diff
@@ -10,16 +10,20 @@
...
+];
 
 !(isNull _unit ||
     {!(_unit isKindOf "CAManBase")} ||
     {!local _unit} ||
-    {(_target distance _unit) > REARM_ACTION_DISTANCE} ||
-    {!isNull (_unit getVariable [QGVAR(dummy), objNull])})
+    {(_truck distance _unit) > REARM_ACTION_DISTANCE} ||
+    {!isNull (_unit getVariable [QGVAR(dummy), objNull])} ||
+    {GVAR(level) == 0})
```
Comment (jonpas) [inline addons/rearm/functions/fnc_canTakeAmmo.sqf]: No indentation needed either.

File: addons/common/functions/fnc_displayTextStructured.sqf (hunk 2 above)
Comment (jonpas) [inline addons/common/functions/fnc_displayTextStructured.sqf]: I guess it doesn't hurt but it will make it inconsistent with other hints. @commy2 ?

File: addons/rearm/functions/fnc_dropAmmo.sqf [hunk 17]
```diff
@@ -4,8 +4,8 @@
  *
  * Arguments:
  * 0: Unit <OBJECT>
- * 1: Delete dummy object <BOOL> (optional)
- * 2: Unholster Weapon <BOOL> (optional)
+ * 1: Delete dummy object <BOOL><OPTIONAL>
+ * 2: Unholster Weapon <BOOL><OPTIONAL>
```
Comment (jonpas) [inline addons/rearm/functions/fnc_dropAmmo.sqf]: Wrong optional argument style.

File: addons/rearm/functions/fnc_getCaliber.sqf [hunk 18]
```diff
@@ -0,0 +1,50 @@
+/*
+ * Author: GitHawk
+ * Get the caliber of the ammo in a magazine and return its parameters.
+ *
+ * Argument:
```
Comment (jonpas) [inline addons/rearm/functions/fnc_getCaliber.sqf]: Argument**s**

File: addons/rearm/functions/fnc_getCaliber.sqf [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+ * 0: Magazine Classname <STRING>
+ *
+ * Return value:
```
Comment (jonpas) [inline addons/rearm/functions/fnc_getCaliber.sqf]: **V**alue

File: addons/rearm/functions/fnc_getSupplyCount.sqf [hunk 20]
```diff
@@ -0,0 +1,29 @@
+/*
+ * Author: GitHawk
+ * Get the supply count.
+ *
+ * Argument:
```
Comment (jonpas) [inline addons/rearm/functions/fnc_getSupplyCount.sqf]: Argument**s**

File: addons/rearm/functions/fnc_getSupplyCount.sqf [hunk 21]
```diff
@@ -0,0 +1,29 @@
...
+ * 0: Ammo Truck <OBJECT>
+ *
+ * Return value:
```
Comment (jonpas) [inline addons/rearm/functions/fnc_getSupplyCount.sqf]: **V**alue

File: addons/rearm/functions/fnc_hasEnoughSupply.sqf [hunk 22]
```diff
@@ -0,0 +1,42 @@
+/*
+ * Author: GitHawk
+ * Check whether enough supply is left to take the magazine.
+ *
+ * Argument:
```
Comment (jonpas) [inline addons/rearm/functions/fnc_hasEnoughSupply.sqf]: Argument**s**

File: addons/rearm/functions/fnc_hasEnoughSupply.sqf [hunk 23]
```diff
@@ -0,0 +1,42 @@
...
+ * 0: Ammo Truck <OBJECT>
+ * 1: Magazine Classname <STRING>
+ *
+ * Return value:
```
Comment (jonpas) [inline addons/rearm/functions/fnc_hasEnoughSupply.sqf]: **V**alue

File: addons/rearm/functions/fnc_hasEnoughSupply.sqf [hunk 24]
```diff
@@ -0,0 +1,42 @@
...
+ */
+#include "script_component.hpp"
+
+params [
+    ["_truck", objNull, [objNull]],
+    ["_magazineClass", "", [""]]
+];
+
+if (isNull _truck ||
+    {_magazineClass isEqualTo ""}) exitWith {false};
```
Comment (jonpas) [inline addons/rearm/functions/fnc_hasEnoughSupply.sqf]: Would look nicer on same line.

File: addons/rearm/functions/fnc_magazineInSupply.sqf [hunk 25]
```diff
@@ -0,0 +1,42 @@
+/*
+ * Author: GitHawk
+ * Returns true if the magazine is in the current supply
+ *
+ * Argument:
```
Comment (jonpas) [inline addons/rearm/functions/fnc_magazineInSupply.sqf]: s

File: addons/rearm/functions/fnc_magazineInSupply.sqf [hunk 26]
```diff
@@ -0,0 +1,42 @@
...
+ * 0: Ammo Truck <OBJECT>
+ * 1: Magazine Classname <STRING>
+ *
+ * Return value:
```
Comment (jonpas) [inline addons/rearm/functions/fnc_magazineInSupply.sqf]: V

File: addons/rearm/functions/fnc_magazineInSupply.sqf [hunk 27]
```diff
@@ -0,0 +1,42 @@
...
+ */
+#include "script_component.hpp"
+
+params [
+    ["_truck", objNull, [objNull]],
+    ["_magazineClass", "", [""]]
+];
+
+if (isNull _truck ||
+    {_magazineClass isEqualTo ""}) exitWith {false};
```
Comment (jonpas) [inline addons/rearm/functions/fnc_magazineInSupply.sqf]: Same line.

File: addons/rearm/functions/fnc_magazineInSupply.sqf [hunk 28]
```diff
@@ -0,0 +1,42 @@
...
+    private _magazinePresent = false;
+    {
+        _x params ["_magazine", "_rounds"];
+        if ((_magazine isEqualTo _magazineClass) && (_rounds > 0)) exitWith {_magazinePresent = true; };
+        false
+    } count _magazineSupply;
+    _magazinePresent
+};
+
+// With infinite supply, the is always one left
```
Comment (jonpas) [inline addons/rearm/functions/fnc_magazineInSupply.sqf]: there

File: addons/rearm/functions/fnc_removeMagazineFromSupply.sqf [hunk 29]
```diff
@@ -0,0 +1,90 @@
+/*
+ * Author: GitHawk
+ * Removes a magazine from the supply.
+ *
+ * Argument:
```
Comment (jonpas) [inline addons/rearm/functions/fnc_removeMagazineFromSupply.sqf]: s

File: addons/rearm/functions/fnc_removeMagazineFromSupply.sqf [hunk 30]
```diff
@@ -0,0 +1,90 @@
...
+ * 0: Ammo Truck <OBJECT>
+ * 1: Magazine Classname <STRING>
+ * 2: Number of Rounds to withdraw <NUMBER><OPTIONAL>
+ *
+ * Return value:
```
Comment (jonpas) [inline addons/rearm/functions/fnc_removeMagazineFromSupply.sqf]: V

File: addons/rearm/functions/fnc_removeMagazineFromSupply.sqf [hunk 31]
```diff
@@ -0,0 +1,90 @@
...
+#include "script_component.hpp"
+
+params [
+    ["_truck", objNull, [objNull]],
+    ["_magazineClass", "", [""]],
+    ["_numRounds", -1, [0]]
+];
+
+if (isNull _truck ||
+    {_magazineClass isEqualTo ""}) exitWith {false};
```
Comment (jonpas) [inline addons/rearm/functions/fnc_removeMagazineFromSupply.sqf]: Same line.

File: addons/rearm/functions/fnc_setSupplyCount.sqf [hunk 32]
```diff
@@ -0,0 +1,26 @@
+/*
+ * Author: GitHawk
+ * Sets the supply count.
+ *
+ * Argument:
```
Comment (jonpas) [inline addons/rearm/functions/fnc_setSupplyCount.sqf]: s

File: addons/rearm/functions/fnc_setSupplyCount.sqf [hunk 33]
```diff
@@ -0,0 +1,26 @@
...
+ * 0: Ammo Truck <OBJECT>
+ * 1: Supply Count <NUMBER>
+ *
+ * Return value:
```
Comment (jonpas) [inline addons/rearm/functions/fnc_setSupplyCount.sqf]: V

File: addons/rearm/CfgVehicles.hpp (hunk 3 above)
Comment (GitHawk) [inline addons/rearm/CfgVehicles.hpp]: Nothing moved, just GH f''''ing up

File: addons/rearm/functions/fnc_addMagazineToSupply.sqf (hunk 4 above)
Comment (GitHawk) [inline addons/rearm/functions/fnc_addMagazineToSupply.sqf]: Well then I look at the wrong reference which used this style.

File: addons/rearm/functions/fnc_canReadSupplyCounter.sqf (hunk 13 above)
Comment (GitHawk) [inline addons/rearm/functions/fnc_canReadSupplyCounter.sqf]: I think it does, since we haven't left the `(`

File: addons/rearm/functions/fnc_dropAmmo.sqf (hunk 17 above)
Comment (GitHawk) [inline addons/rearm/functions/fnc_dropAmmo.sqf]: Wrong reference used again.

File: addons/common/functions/fnc_displayTextStructured.sqf (hunk 2 above)
Comment (GitHawk) [inline addons/common/functions/fnc_displayTextStructured.sqf]: No it won't. The default value is the old fixed value and since it's optional nothing else is broken.

**Comment by GitHawk** [comment]
Since the first merge broke a lot of stuff, everything should be fine now.

**Comment by bux578** [comment]
Why is this branch both change of headers and the rearm modification?
Imo it should be two separate branches, to better keep track of the changes.
Would this be possible to do for you @GitHawk

**Comment by GitHawk** [comment]
Unfortunately no. I made this PR in February. In June it gets reviewed. Some changed to headers are suggested, as they changed since. To avoid me and others from making the same (copy paste) error again and again, I replaced all headers that were "wrong".

Also I'm getting sick and tired of keeping my month-old PRs up to date. If you want some changes made, please PR me on my branch.

**Comment by jonpas** [comment]
We have priorities, and this is a large change in code
Additionally this can't get merged before change to CBA events is, so it will be even more time before we'll be able to merge it. Sorry for inconvenience.

**Comment by jonpas** [comment]
Will re-PR this in multiple PRs appropriately.

**Comment by jonpas** [comment]
Separated into #3943 and #3944.

I won't be doing API things, if you can please re-PR that @GitHawk.
